### PR TITLE
sdxl: add timestep bias strategies

### PR DESCRIPTION
**Pull request by Fzilan**
add timestep bias strategies when training.

### CODE REVIEW

**Review by zhanghuiyao — APPROVED** [review]

**Review by geniuspatrick — APPROVED** [review]

**Review by SamitHuang — APPROVED** [review]

**Comment by SamitHuang** [comment]
@Fzilan  Please also update the doc to add instructions for enabling timestep bias strategy

File: examples/stable_diffusion_xl/gm/models/trainer_factory.py [hunk 1]
```diff
@@ -64,14 +70,65 @@ def __init__(
...
+                )
+            if range_end > num_timesteps:
+                raise ValueError(
+                    "When using the range strategy for timestep bias, you must provide an ending timestep smaller than the number of timesteps."
+                )
+            bias_indices = slice(range_begin, range_end)
+        else:  # 'none' or any other string
+            return weights
+        if self.timestep_bias_multiplier <= 0:
+            return ValueError(
```
Comment (hadipash) [inline examples/stable_diffusion_xl/gm/models/trainer_factory.py:106]: raise

File: examples/stable_diffusion_xl/gm/models/trainer_factory.py [hunk 2]
```diff
@@ -64,14 +70,65 @@ def __init__(
...
+
     def construct(self, x, *tokens):
         # get latent target
         if self.enable_first_stage_model:
             x = self.first_stage_model.encode(x)
         x = self.scale_factor * x
 
         # get noise and sigma
-        sigmas = self.sigma_sampler(x.shape[0])
+        if self.timestep_bias_strategy == "none":
```
Comment (hadipash) [inline examples/stable_diffusion_xl/gm/models/trainer_factory.py:127]: Why `"none"` and not `None`?

File: examples/stable_diffusion_xl/gm/models/trainer_factory.py [hunk 3]
```diff
@@ -64,14 +70,65 @@ def __init__(
...
         self.enable_first_stage_model = enable_first_stage_model
 
+        self.timestep_bias_strategy = timestep_bias_strategy
+        self.timestep_bias_multiplier = timestep_bias_multiplier
+        self.timestep_bias_begin = timestep_bias_begin
+        self.timestep_bias_end = timestep_bias_end
+        self.timestep_bias_portion = timestep_bias_portion
+        self.timestep_bias_weighting = Tensor(self.generate_timestep_weights(model.sigma_sampler.num_idx))
+
+    def generate_timestep_weights(self, num_timesteps):
```
Comment (hadipash) [inline examples/stable_diffusion_xl/gm/models/trainer_factory.py:80]: The generation of `timestep_bias_weighting` can be extracted from `TrainOneStepCell`, and only a single `timestep_bias_weighting` variable needs to be passed to `TrainOneStepCell`. Avoid overcluttering `TrainOneStepCell`.

**Comment by SamitHuang** [comment]
Please fix the conflicts and comments

File: examples/stable_diffusion_xl/gm/models/trainer_factory.py (hunk 2 above)
Comment (Fzilan) [reply examples/stable_diffusion_xl/gm/models/trainer_factory.py:127]: the type of arg "timestep_bias_strategy" set to str and choose from ["earlier", "later", "range", "none"]

Comment (hadipash) [reply examples/stable_diffusion_xl/gm/models/trainer_factory.py:127]: remove `default="none"` from the parser and the value will be `None` by default when no arguments are specified. Don't use `"none"` as a substitution of `None`.

**Pull request merged**